File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/planning/ProviderCostSelectionPolicy.java
package com.foundgine.core.semantic.planning;

/** Controls how provider execution cost participates in rewrite ranking. */
public record ProviderCostSelectionPolicy(boolean preferLowerProviderCost,double providerCostWeight) {
    public ProviderCostSelectionPolicy(){this(true,1d);}
    public ProviderCostSelectionPolicy validate(){if(Double.isNaN(providerCostWeight)||Double.isInfinite(providerCostWeight)||providerCostWeight<0)throw new IllegalArgumentException("Provider cost weight must be finite and non-negative.");return this;}
}

File: src/csharp/Foundgine.Core/Semantic/Planning/ProviderCostSelectionPolicy.cs
namespace Foundgine.Core.Semantic.Planning;

/// <summary>Controls how provider execution cost participates in rewrite ranking.</summary>
public sealed record ProviderCostSelectionPolicy(
    bool PreferLowerProviderCost = true,
    double ProviderCostWeight = 1d)
{
    public ProviderCostSelectionPolicy Validate()
    {
        if (double.IsNaN(ProviderCostWeight) || double.IsInfinity(ProviderCostWeight) || ProviderCostWeight < 0d)
            throw new ArgumentOutOfRangeException(nameof(ProviderCostWeight));
        return this;
    }
}
